### PR TITLE
fix: getCallFrame does not need to throw

### LINKED ISSUE #562: getCallFrame does not need to throw + test in other browsers
## Describe the bug

As discussed in live stream, https://github.com/mswjs/msw/blob/5b232eb181c9b78bc0fb26954b448e5ccd26ed24/src/utils/internal/getCallFrame.ts does not need the Error to be thrown, just `new Error().stack` is enough.

---

The difference between browsers should not an issue after all, but FTR: Chrome and Node, unlike Firefox and Safari, contain the error message on the 1st line of stack (but the fist line in Firefox/Safari will be `getCallFrame` itself, so doing `frames.slice(1)` is safe). In any case, probably best to test in multiple browsers just in case.

Executing `(function a() { console.log(new Error().stack) })()` in Chrome:
```
Error
    at a (<anonymous>:1:29)
    at <anonymous>:1:50
```

in Firefox:
```
a@debugger eval code:1:28
@debugger eval code:1:1
```

### PR DESCRIPTION
**Pull request by Aprillion**
Fixes #562

### CODE REVIEW

File: src/utils/internal/getCallFrame.ts [hunk 1]
```diff
@@ -2,25 +2,21 @@
...
 export function getCallFrame() {
-  try {
-    const inspectionError = new Error()
-    inspectionError.name = 'Inspection Error'
-    throw inspectionError
-  } catch (error) {
-    const frames: string[] = error.stack.split('\n')
+  // Error.stack is available even in IE 10, no need for nullish coalesce
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Stack#browser_compatibility
+  const frames: string[] = (new Error().stack as string).split('\n')
```
Comment (timdeschryver) [inline src/utils/internal/getCallFrame.ts:5]: Nit, since `stack` is a string (or can be nullable) I prefer the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator) instead of a cast.

```suggestion
  const frames: string[] = new Error().stack!.split('\n')
```

**Review by timdeschryver — DISMISSED** [review]
👍👍

File: src/utils/internal/getCallFrame.ts (hunk 1 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts:5]: done... 

later undone because:
a) my local git somehow did not fetch the commit made in github :(
b) `!` is forbidden by eslint in this repo

File: src/utils/internal/getCallFrame.ts [hunk 2]
```diff
@@ -2,25 +2,21 @@
  * Return the stack trace frame of a function's invocation.
  */
 export function getCallFrame() {
-  try {
-    const inspectionError = new Error()
-    inspectionError.name = 'Inspection Error'
-    throw inspectionError
-  } catch (error) {
-    const frames: string[] = error.stack.split('\n')
+  // Error.stack is available even in IE 10, no need for nullish coalesce
```
Comment (kettanaito) [inline src/utils/internal/getCallFrame.ts]: I don't think this comment is necessary. At least, I wouldn't reference MDN link.

File: src/utils/internal/getCallFrame.ts [hunk 3]
```diff
@@ -2,25 +2,21 @@
...
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Stack#browser_compatibility
+  const frames: string[] = new Error().stack!.split('\n')
```
Comment (kettanaito) [inline src/utils/internal/getCallFrame.ts]: Do we need `!` here because `Error.stack` is typed as optional in TypeScript?

File: src/utils/internal/getCallFrame.ts [hunk 4]
```diff
@@ -2,25 +2,21 @@
...
-
-    // Extract file reference from the stack frame.
-    const [, declarationPath] = declarationFrame.match(/\((.+?)\)$/) || []
-    return declarationPath
+  if (!declarationFrame) {
+    return
   }
+
+  // Extract file reference from the stack frame.
+  const [, declarationPath] = declarationFrame.match(/\((.+?)\)$/) || []
```
Comment (kettanaito) [inline src/utils/internal/getCallFrame.ts]: Did you have a chance to examine whether Firefox stack traces match this format? 

```
  at (<FILE_PATH>)
```

**Review by kettanaito — CHANGES_REQUESTED** [review]
Thank you for the improvement, @Aprillion! I've left a few comments, mainly questions to understand the changes.

File: src/utils/internal/getCallFrame.ts (hunk 2 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: ok, I can remove both comment lines... (probably not today, still catching up with day-job work)

File: src/utils/internal/getCallFrame.ts (hunk 3 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: yes. TS definitions for web apis are quite out of date TBH

P.S.: I don't like that `Resolve conversation` button deletes my draft reply

File: src/utils/internal/getCallFrame.ts (hunk 4 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: nope. do you have any instructions how to execute this code in browser? I can click the `here` button from CodeSandbox comment above, but what can I do to execute `getCallFrame` plz???

File: src/utils/internal/getCallFrame.ts (hunk 2 above)
Comment (kettanaito) [reply src/utils/internal/getCallFrame.ts]: No pressure. I'm thankful for your contribution already!

File: src/utils/internal/getCallFrame.ts (hunk 4 above)
Comment (kettanaito) [reply src/utils/internal/getCallFrame.ts]: Sure, I'll share the instructions below. 

## Executing in a browser

1. Clone the repo and install its dependencies.
2. Run a single integration test in a browser. You'd need a test that exposes the `msw` object globally. For example `yarn test:focused test/msw-api/setup-worker/printHandlers.mocks.ts`
3. Open the spawned URL in Firefox.
4. Open DevTools, switch to "Console".
5. Run `window.msw.worker.printHandlers()`.

That should print the list of handlers into the console. As a part of each message, there will be a call frame to the handler's definition.

Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: Screenshot from output in Chrome (left), Safari (middle), and Firefox (right),, on macOS 10.14:

Screenshot 2021-01-29 at 19 11 22

Looks like the "Declaration" property is `undefined` => I will try to investigate.
(edit: replaced screenshot to show the problem)

File: src/utils/internal/getCallFrame.ts (hunk 2 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: now removed

File: src/utils/internal/getCallFrame.ts (hunk 3 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: looks like you guys disabled `!` in eslint => I am returning back to my original `as string` for now

Screenshot 2021-01-29 at 19 46 06

File: src/utils/internal/getCallFrame.ts (hunk 4 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: OK, I pushed a commit that will make Firefox output the same Declaration as Chrome, and Safari to output the best it can.

The unit test is necessary for TDD. As I mentioned in issue #562, I was not able to run the integration / focused test on my Windows => if someone else can add unit tests from more supported browsers or from a Deno server, feel free to tag me to help with improving the implementation.

Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: arghh@!$# I need to fix `setup-server/printHandlers.test.ts` .. looks like Node has a slightly different stack than Chrome after all

edit: fixed

**Comment by Aprillion** [comment]
Hmm, this will need more testing on Windows, but please review the unit test approach so far - whether it makes sense for me to continue with this approach (with my luck so far, the amount of edge cases from Windows will double the amount of unit tests).

**Comment by Aprillion** [comment]
Tests on Windows are now green too. 🥳

File: src/utils/internal/getCallFrame.test.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
+import { getCallFrame } from './getCallFrame'
```
Comment (kettanaito) [inline src/utils/internal/getCallFrame.test.ts:1]: This test looks great! 
We should also have the same test suite for NodeJS. Stack traces there may differ, it would be nice to test them.

```js
// src/utils/internal/getStackFrame.node.test.ts
/**
 * @jest-environment node
 */
```

Would you be interested in adding the NodeJS tests? If you don't have time for this I can give you a hand.

File: src/utils/internal/getCallFrame.test.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+const mockStack = jest.fn()
+jest.spyOn(window, 'Error').mockImplementation(() => ({
```
Comment (kettanaito) [inline src/utils/internal/getCallFrame.test.ts]: Please put this spying in the `beforeAll` hook. Don't forget to add the `afterAll` hook that `jest.restoreAllMocks()` clears things out.

File: src/utils/internal/getCallFrame.test.ts (hunk 5 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.test.ts:1]: Hm, this is just a unit test with a mock for `new Error` ... using fake stack traces to simulate NodeJS from both macOS and Windows.

And a real native NodeJS `new Error().stack` is used in the integration test `setup-server/printHandlers.test.ts` already. IMHO no need to test mock errors in 2 different jest environments..

File: src/utils/internal/getCallFrame.test.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+      '    at getCallFrame (/Users/mock/github/msw/lib/umd/index.js:3735:22)',
+      '    at Object.get (/Users/mock/github/msw/lib/umd/index.js:3776:29)',
+      '    at Object.<anonymous> (/Users/mock/github/msw/test/msw-api/setup-server/printHandlers.test.ts:13:8)', // <-- this one
+      '    at Runtime._execModule (/Users/mock/github/msw/node_modules/jest-runtime/build/index.js:1299:24)',
+      '    at Runtime._loadModule (/Users/mock/github/msw/node_modules/jest-runtime/build/index.js:898:12)',
+      '    at Runtime.requireModule (/Users/mock/github/msw/node_modules/jest-runtime/build/index.js:746:10)',
+      '    at jasmine2 (/Users/mock/github/msw/node_modules/jest-jasmine2/build/index.js:230:13)',
+      '    at runTestInternal (/Users/mock/github/msw/node_modules/jest-runner/build/runTest.js:380:22)',
+      '    at runTest (/Users/mock/github/msw/node_modules/jest-runner/build/runTest.js:472:34)',
+    ].join('\n'),
```
Comment (kettanaito) [inline src/utils/internal/getCallFrame.test.ts:31]: Alternatively, we could return a template string literal that preserves line breaks. Do you think that would be different from the current approach?

File: src/utils/internal/getCallFrame.ts [hunk 8]
```diff
@@ -2,25 +2,22 @@
...
+  const frames: string[] = (new Error().stack as string).split('\n')
 
-    // Get the first frame that doesn't reference the library's internal trace.
-    // Assume that frame is the invocation frame.
-    const declarationFrame = frames.slice(1).find((frame) => {
-      return !/(node_modules)?\/lib\/(umd|esm)\//.test(frame)
-    })
+  // Get the first frame that doesn't reference the library's internal trace.
+  // Assume that frame is the invocation frame.
+  const rePathIsNotUseful = /(node_modules)?[\/\\]lib[\/\\](umd|esm)[\/\\]|^[^\/\\]*$/
```
Comment (kettanaito) [inline src/utils/internal/getCallFrame.ts]: I was thinking about utilizing `path.resolve` and its built-in system-specific path delimiters. I suppose that won't work in a browser, will it?

Comment (kettanaito) [inline src/utils/internal/getCallFrame.ts]: So `node_modules` part is not necessary? Interesting. Could you share more info when does this happen?

Comment (kettanaito) [inline src/utils/internal/getCallFrame.ts]: What do you think if we rename this variable? 
```suggestion
  const ignoreFrameRegExp= /(node_modules)?[\/\\]lib[\/\\](umd|esm)[\/\\]|^[^\/\\]*$/
```

File: src/utils/internal/getCallFrame.test.ts (hunk 5 above)
Comment (kettanaito) [reply src/utils/internal/getCallFrame.test.ts:1]: Oh, I see. Since we are using an exact system-specific stack trace there's no need to re-create this test suite in a different environment. Nice catch!

File: src/utils/internal/getCallFrame.test.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+      '    at runTest (/Users/mock/git/msw/node_modules/jest-runner/build/runTest.js:472:34)',
+    ].join('\n'),
+  )
+
+  expect(getCallFrame()).toBe(
+    '/Users/mock/git/msw/test/msw-api/setup-server/printHandlers.test.ts:14:11',
+  )
+})
+
+test('Node on Windows error stack', () => {
```
Comment (kettanaito) [inline src/utils/internal/getCallFrame.test.ts:60]: Do you mind grouping these test scenarios under their respective OS?

```js
describe('MacOS', () => {
  // ...
})

describe('Windows', () => {
  // ...
//
```

I agree that flat tests are good, but in this case grouping them would help scanning trough quicker.

**Review by kettanaito — CHANGES_REQUESTED** [review]
I've left a few suggestions and a question. Would appreciate your help.

File: src/utils/internal/getCallFrame.test.ts (hunk 6 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.test.ts]: I thought that Jest spies work on the level of a single test file, but can't any documentation whether that's true or not :(

- [x] add explicit beforeAll/afterAll

File: src/utils/internal/getCallFrame.test.ts (hunk 7 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.test.ts:31]: I added `console.warn(frames)` to the function and copy&pasted the arrays from console :) Otherwise the only difference would be in indentation.

Comment (Aprillion) [reply src/utils/internal/getCallFrame.test.ts:31]: oh, and I couldn't do the `// <-- this one` comment .. I used it only to guide me to make the assertion, not sure how useful it is to keep those comments for future debugging 🤷‍♂️

=> not changing now, but if you tell me you really really prefer template string literal without indentation and without line comments, I can convert.

File: src/utils/internal/getCallFrame.ts (hunk 8 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: yeah, browsers use `webpack:///./test/msw-api/setup-worker/printHandlers.mocks.ts?:6:113` kind of path on all OS

Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: I didn't change anything about that part, `(node_modules)?` was there already in commit 944d805

Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: - [x] rename to `ignoreFrameRegExp`

File: src/utils/internal/getCallFrame.test.ts (hunk 9 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.test.ts:53]: - [ ] group tests by OS

Comment (Aprillion) [reply src/utils/internal/getCallFrame.test.ts:60]: hm,, what do you want me to do with `Chrome and Edge error stack` and `Firefox on macOS and Windows error stack` tests please? they have the same error stack on both

File: src/utils/internal/getCallFrame.ts (hunk 8 above)
Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: But thinking about it, it probably wouldn't make any difference if we remove that part from the regex, `/(node_modules)?[\/\\]lib[\/\\](umd|esm)[\/\\]|^[^\/\\]*$/` wouldn't match anything more than `/[\/\\]lib[\/\\](umd|esm)[\/\\]|^[^\/\\]*$/` would... I think 🤔

Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: Double but, what if someone will use `/lib/esm/msw.js` file structure to set up the handlers in their repo?  While we wouldn't want to ignore that and in their case `msw` library would be inside node_modules indeed .. but I cannot think of a simple way how to make `getCallFrame` testable inside this repo as well as handle such a special edge case when msw is used from node_modules.

One idea was to detect `msw/lib` in the path to skip those frames, but there are both false positives (someone using "msw" as a directory name to set up msw handlers) and false negatives (someone cloning this repo into a folder different from "msw").

Unless someone has a better idea for more reliable detection of "internal msw" vs "user space", IMHO let's ignore that for now 😰

File: src/utils/internal/getCallFrame.ts [hunk 10]
```diff
@@ -2,25 +2,22 @@
...
+  const frames: string[] = (new Error().stack as string).split('\n')
 
-    // Get the first frame that doesn't reference the library's internal trace.
-    // Assume that frame is the invocation frame.
-    const declarationFrame = frames.slice(1).find((frame) => {
-      return !/(node_modules)?\/lib\/(umd|esm)\//.test(frame)
-    })
+  // Get the first frame that doesn't reference the library's internal trace.
+  // Assume that frame is the invocation frame.
+  const ignoreFrameRegExp = /(node_modules)?[\/\\]lib[\/\\](umd|esm)[\/\\]|^[^\/\\]*$/
```
Comment (timdeschryver) [inline src/utils/internal/getCallFrame.ts]: I might be wrong here, but shouldn't we also at the `iife` build here?
```suggestion
  const ignoreFrameRegExp = /(node_modules)?[\/\\]lib[\/\\](umd|esm|iife)[\/\\]|^[^\/\\]*$/
```

File: src/utils/internal/getCallFrame.test.ts (hunk 9 above)
Comment (kettanaito) [reply src/utils/internal/getCallFrame.test.ts:60]: Oh, a good point. I haven't noticed some of the test suites shared the stack trace across OS. I don't mind, let's keep it as-is.

File: src/utils/internal/getCallFrame.ts (hunk 10 above)
Comment (kettanaito) [reply src/utils/internal/getCallFrame.ts]: @timdeschryver I think you're totally right. Ideally, we should include all the build targets (umd, esm, cjs, iife) so all source file references can be accounted for.

Comment (Aprillion) [reply src/utils/internal/getCallFrame.ts]: - [x] add `|iife|cjs` to regex

**Review by kettanaito — APPROVED** [review]
Great job, @Aprillion! Welcome to contributors!
